File: app/models/game.rb
class Game
  class << self
    def start user, opponent_user
      white, black = [user, opponent_user].shuffle
      ActionCable.server.broadcast "player_#{white}",
        {action: "game_start", msg: "white"}
      ActionCable.server.broadcast "player_#{black}",
        {action: "game_start", msg: "black"}

      REDIS.set("opponent_for:#{white}", black)
      REDIS.set("opponent_for:#{black}", white)
    end

    def forfeit user
      if winner = opponent_for(user)
        ActionCable.server.broadcast "player_#{winner}",
          {action: "opponent_forfeits", msg: "Opponent disconnected, you win!"}
      end
    end

    def opponent_for user
      REDIS.get("opponent_for:#{user}")
    end

    def make_move user, data
      opponent = opponent_for user
      move_string = "#{data["move"]["from"]}-#{data["move"]["to"]}"
      status = data["status"].split("-")[0]
      type = data["status"].split("-")[1]
      ActionCable.server.broadcast "player_#{opponent}",
        {action: "make_move", move: move_string, msg: status, type: type}
    end

    def game_over user
      opponent = opponent_for user
      ActionCable.server.broadcast "player_#{opponent}",
        {action: "game_over", msg: "You lose!"}
    end
  end
end

File: app/assets/javascripts/channels/game.js
App.game = App.cable.subscriptions.create("GameChannel", {
  connected: function() {
    return this.printMessage("Waiting for opponent", "wait-opponent");
  },
  received: function(data) {
    var ref, source, target;
    switch (data.action) {
      case "game_start":
        App.board.position("start");
        App.board.orientation(data.msg);
        return this.printMessage("Game started! You play as " + data.msg + ".", "notice");
      case "make_move":
        ref = data.move.split("-"), source = ref[0], target = ref[1];
        App.chess.move({
          from: source,
          to: target,
          promotion: "q"
        });
        this.printMessage(data.msg, data.type);
        return App.board.position(App.chess.fen());
      case "opponent_forfeits":
        return this.printMessage(data.msg, data.type);
      case "game_over":
        return this.printMessage(data.msg, data.type);
    }
  },
  printMessage: function(message, type) {
    if(type === "wait-opponent" && $("#game-status").children().length === 0) {
      $("#game-status").append("<p class=" + type + ">" + message + "</p>");
    }
    if (type !== "wait-opponent"){
      $("#game-status").empty();
      $("#game-status").append("<p class=" + type + ">" + message + "</p>");
    }
  }
});
